hal/hal-esp-idf.c

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: Apache-2.0
+#include <hal.h>
+#include <hal/systimer_hal.h>
+#include <hal/wdt_hal.h>
+#include <hal/cache_ll.h>
+#include <hal/cache_hal.h>
+#include <soc/soc.h>
+#include <esp_private/systimer.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdio.h>
+
+extern int main();
+
+/*NOTE: refer to esp-idf/components/bootloader/subproject/main/ld/esp32c3/bootloader.ld*/
+extern char _stack_bottom; /* start of the stack */
+static char *stack_start = &_stack_bottom;
+static systimer_hal_context_t hal_ctx;
+
+extern int uart_tx_one_char(uint8_t c);
+void hal_setup(const enum clock_mode clock) {
+    wdt_hal_context_t wdt0_context = {.inst = WDT_MWDT0, .mwdt_dev = &TIMERG0};
+    wdt_hal_write_protect_disable(&wdt0_context);
+    wdt_hal_disable(&wdt0_context);
+    wdt_hal_set_flashboot_en(&wdt0_context, false);
+
+    wdt_hal_context_t rwdt_context = RWDT_HAL_CONTEXT_DEFAULT();
+    wdt_hal_write_protect_disable(&rwdt_context);
+    wdt_hal_disable(&rwdt_context);
+    wdt_hal_set_flashboot_en(&rwdt_context, false);
+
+    REG_WRITE(RTC_CNTL_SWD_WPROTECT_REG, RTC_CNTL_SWD_WKEY_VALUE);
+    REG_SET_BIT(RTC_CNTL_SWD_CONF_REG, RTC_CNTL_SWD_AUTO_FEED_EN);
+    REG_WRITE(RTC_CNTL_SWD_WPROTECT_REG, 0);
+
+    cache_hal_init();
+    cache_hal_is_cache_enabled(CACHE_LL_ID_ALL, CACHE_TYPE_ALL);
+
+    (void)clock;
+    systimer_hal_init(&hal_ctx);
+    systimer_hal_tick_rate_ops_t ops = {
+        .ticks_to_us = systimer_ticks_to_us,
+        .us_to_ticks = systimer_us_to_ticks,
+    };
+
+    systimer_hal_set_tick_rate_ops(&hal_ctx, &ops);
+    systimer_hal_enable_counter(&hal_ctx, 0);
```

Comment:
some comments on what this is actually doing may be helpful if we ever have to look at this again.